dv/formal/conductor.py

```diff
@@ -0,0 +1,802 @@
+# counterexamples, and analyse the results.
+
+import asyncio
+import argparse
+import re
+import subprocess
+import time
+import json
+import os
+from math import floor
+from datetime import datetime
+import hashlib
+
+# --------------------- Utilities ----------------------
+
+PROOFLOG = os.environ["PROOFLOG"] if "PROOFLOG" in os.environ else "prooflog.txt"
+LOGFILE = os.environ["LOGFILE"] if "LOGFILE" in os.environ else "logfile.txt"
+
+# Send all printed text to a log file, and prepend the date and time.
+# This is helpful for looking back at how long things took, or how
+# long until we expect something to be done.
+oldprint = print
+def print(*args):
+    timestr = datetime.now().strftime('[%d/%m/%y %H:%M:%S.%f]')
+    oldprint(gray(timestr), *args)
+    if not "NO_LOG" in os.environ:
+        with open(LOGFILE, "a") as f:
+            oldprint(timestr, *[re.sub("\033\\[[0-9;]+m", "", str(x)) for x in args], file=f)
+if not "NO_LOG" in os.environ:
+    with open(LOGFILE, "a") as f:
+        f.write("\n")
+
+def green(s): return f"\033[32;1m{s}\033[0m"
+def red(s): return f"\033[31;1m{s}\033[0m"
+def white(s): return f"\033[1m{s}\033[0m"
+def orange(s): return f"\033[33;1m{s}\033[0m"
+def gray(s): return f"\033[90m{s}\033[0m"
+
+# Races allow async processes to cleanly run in parrallel, until one declares victory, then
+# all the other threads are cancelled.
+class RaceWonException(Exception):
+    def __init__(self, result):
+        super().__init__()
+        self.result = result
+
+def won_race(result):
+    raise RaceWonException(result)
+
+async def race(tasks):
+    winner = None
+    try:
+        async with asyncio.TaskGroup() as tg:
+            for task in tasks:
+                tg.create_task(task)
+    except* RaceWonException as e:
+        for result in e.exceptions:
+            winner = result.result
+    return winner
+
+# --------------------- Process Management ----------------------
+
+# Returns the global free memory according to /proc/meminfo, note though that reclaimable
+# memory (e.g. cached files) are not added to this when they could be, leading to the impression
+# of less available memory than there is for our use cases.
+def global_memory_free():
+    with open("/proc/meminfo", "r") as f:
+        c = f.read()
+        for line in c.split("\n"):
+            if line.startswith("MemFree:"):
+                return float(line[8:-2].strip()) / (1024 * 1024)
+    return 0
+
+# Memory used by a specific PID according to /proc/{pid}/status
+def pid_memory_used(pid):
+    with open(f"/proc/{pid}/status", "r") as f:
+        c = f.read()
+        for line in c.split("\n"):
+            if line.startswith("RssAnon:"):
+                return float(line[8:-2].strip()) / (1024 * 1024)
+    return 0
+
+# Finishes when a process exits, and cancelling kills it.
+# Result is (exit code, max mem, dt, status string, stdout?, stderr?)
```

Comment:
Can we encode these into the class type so that we don't have the confusing type signature later in the Process.__init__ function?